include/ginkgo/core/base/batch_multi_vector.hpp

```diff
@@ -0,0 +1,800 @@
+                                 const BatchMultiVector<ValueType>* b);
+
+    /**
+     * @copydoc compute_dot(const BatchMultiVector *, BatchMultiVector *) const
+     *
+     * @note  Other implementations of batch_multi_vector should override this
+     * function instead of compute_dot(const BatchMultiVector *b,
+     * BatchMultiVector *result).
+     */
+    virtual void compute_dot_impl(const BatchMultiVector<ValueType>* b,
+                                  BatchMultiVector<ValueType>* result) const;
+
+    /**
+     * @copydoc compute_norm2(BatchMultiVector *) const
+     *
+     * @note  Other implementations of batch_multi_vector should override this
+     * function instead of compute_norm2(BatchMultiVector *result).
+     */
+    virtual void compute_norm2_impl(
+        BatchMultiVector<remove_complex<ValueType>>* result) const;
+
+    size_type linearize_index(size_type batch, size_type row,
+                              size_type col) const noexcept
+    {
+        return batch_size_.get_cumulative_offset(batch) +
+               row * batch_size_.get_common_size()[1] + col;
+    }
+
+    size_type linearize_index(size_type batch, size_type idx) const noexcept
+    {
+        return linearize_index(batch, idx / this->get_common_size()[1],
+                               idx % this->get_common_size()[1]);
+    }
+
+private:
+    batch_dim<2> batch_size_;
+    array<value_type> values_;
+};
+
+
+/**
+ * Creates and initializes a batch of column-vectors.
+ *
+ * This function first creates a temporary Dense matrix, fills it with passed in
+ * values, and then converts the vector to the requested type.
+ *
+ * @tparam Matrix  matrix type to initialize
+ *                 (Dense has to implement the ConvertibleTo<Matrix> interface)
+ * @tparam TArgs  argument types for Matrix::create method
+ *                (not including the implied Executor as the first argument)
+ *
+ * @param vals  values used to initialize the batch vector
+ * @param exec  Executor associated to the vector
+ * @param create_args  additional arguments passed to Matrix::create, not
+ *                     including the Executor, which is passed as the first
+ *                     argument
+ *
+ * @ingroup BatchMultiVector
+ * @ingroup mat_formats
+ */
+template <typename Matrix, typename... TArgs>
+std::unique_ptr<Matrix> batch_initialize(
+    std::initializer_list<std::initializer_list<typename Matrix::value_type>>
+        vals,
+    std::shared_ptr<const Executor> exec, TArgs&&... create_args)
+{
+    using batch_multi_vector = BatchMultiVector<typename Matrix::value_type>;
+    size_type num_batch_entries = vals.size();
+    auto vals_begin = begin(vals);
+    size_type common_num_rows = vals_begin->size();
+    auto common_size = dim<2>(common_num_rows, 1);
+    for (size_type b = 0; b < num_batch_entries; ++b) {
+        GKO_ASSERT_EQ(common_num_rows, vals_begin->size());
+        vals_begin++;
+    }
```

Comment:
range based for loop?
```suggestion
    for (const auto& val: vals) {
        GKO_ASSERT_EQ(common_num_rows, val.size());
    }
```
Or perhaps just merge it directly with the loop below.